src/Dscp/Crypto/Encrypt.hs

```diff
@@ -0,0 +1,170 @@
+import Crypto.Error (onCryptoFailure)
+import Data.ByteArray (ByteArray, ByteArrayAccess)
+import qualified Data.ByteArray as BA
+import Data.Text.Buildable (build)
+import Fmt ((+|), (|+))
+import Text.Show (show)
+
+import Dscp.Crypto.Impl (hash)
+
+-------------------------------------------------------------
+-- Passphrases
+-------------------------------------------------------------
+
+-- | Newtype for passphrase. Uses blind instances for
+-- 'Buildable' and 'Show' in order to avoid accidental
+-- appearance of actual passphrases in logs.
+newtype PassPhrase = PassPhrase
+    { getPassPhrase :: ByteString
+    } deriving (Eq, Ord, Monoid, ByteArray, ByteArrayAccess)
+
+instance Buildable PassPhrase where
+    build _ = "<passphrase>"
+
+instance Show PassPhrase where
+    show _ = "<passphrase>"
+
+-- | Minimum passphrase length. Should be enough.
+minPassPhraseLength :: Int
+minPassPhraseLength = 8
+
+-- | Maximum passphrase length. To keep it sane.
+maxPassPhraseLength :: Int
+maxPassPhraseLength = 128
+
+-- | Errors which might happen during passphrase construction.
+data PassPhraseError
+    = PassPhraseTooShort
+        { peLength :: !Int }
+    | PassPhraseTooLong
+        { peLength :: !Int }
+    deriving (Eq)
+
+instance Buildable PassPhraseError where
+    build PassPhraseTooShort {..} =
+        "Passphrase is too short ("+|peLength|+
+        " chars), minimum length is "+|minPassPhraseLength|+" chars."
+    build PassPhraseTooLong {..} =
+        "Passphrase is too long ("+|peLength|+
+        " chars), maximum length is "+|maxPassPhraseLength|+" chars."
+
+instance Show PassPhraseError where
+    show = toString . pretty
+
+instance Exception PassPhraseError
+
+-- | Smart constructor for a passphrase
+mkPassPhrase :: ByteString -> Either PassPhraseError PassPhrase
+mkPassPhrase bs
+    | lbs < minPassPhraseLength = Left $ PassPhraseTooShort lbs
+    | lbs > maxPassPhraseLength = Left $ PassPhraseTooLong lbs
+    | otherwise = Right $ PassPhrase bs
+  where
+    lbs = length bs
+
+-------------------------------------------------------------
+-- Encryption/decryption
+-------------------------------------------------------------
+
+-- | Datatype which combines ciphertext with cipher authentication tag.
+data Encrypted ba = Encrypted
+    { eAuthTag    :: !AuthTag
+    , eCiphertext :: !ba
+    } deriving (Eq)
+
+-- | Authentication tag length. Number 16 is the same as in the
+-- `Crypto.MAC.Poly1305.Auth` tag datatype.
+-- TODO: probably it can be made shorter without any problems?
+authTagLength :: Int
+authTagLength = 16
+
+-- | Chosen cipher. We'll stick with AES256 for now.
+-- TODO: are there better alternatives?
+type CipherType = AES256
+
+-- | Authentication header used for our encryption (empty one).
+authHeader :: ByteString
+authHeader = ""
+
+-- | AEAD mode used for our encryption scheme.
+-- We use GCM mode, because it's the most modern and performant one
```

Comment:
And also because it provides authentication, otherwise we'd had to use some other MAC.